src/main.cpp

```diff
@@ -1,5 +1,130 @@
+#include <Lexer.h>
+#include <Parser.h>
+#include <PrettyPrinter.h>
+
+#include <fstream>
 #include <iostream>
+#include <vector>
+
+// based on https://stackoverflow.com/a/868894
+class ArgParser {
+    std::vector<std::string> args;
+public:
+    enum struct Mode{
+        PrettyPrint,
+        CompileOnly,
+        InterpretOnly,
+        None
+    } mode;
+    std::pair<bool, std::string> optionalFilepath;
+    ArgParser(int argc, char** argv)
+        : mode(Mode::None)
+        , optionalFilepath() {
+        for (int i = 0; i < argc; i++) {
+            args.emplace_back(argv[i]);
+        }
+        if (optionExists("-h") || optionExists("--help")) {
+            printHelpAndExit(0);
+        }
+        setModeIfExistsFlag("-p", Mode::PrettyPrint);
+        setModeIfExistsFlag("-c", Mode::CompileOnly);
+        setModeIfExistsFlag("-b", Mode::InterpretOnly);
+        if (args.back() != "-p" &&
+                args.back() != "-c" &&
+                args.back() != "-b" &&
+                args.back() != args.front()) {
+            optionalFilepath.first = true;
+            optionalFilepath.second = args.back();
+        } else {
+            optionalFilepath.first = false;
+            optionalFilepath.second = "";
+        }
+    }
+private:
+    bool optionExists(const std::string& option) {
+        return std::find(args.begin(), args.end(), option) != args.end();
+    }
+    void setModeIfExistsFlag(const std::string& flag, Mode mode) {
+        if (optionExists(flag)) {
+            if (this->mode != Mode::None) {
+                std::cout << "Cannot specify multiple modes!" << std::endl;
+                printHelpAndExit(1);
+            }
+            this->mode = mode;
+        }
+    }
+    [[noreturn]] void printHelpAndExit(int exitCode) {
```

Comment:
`[[noreturn]]`?

Reply:
Arguably it’s also useful as documrntstion; “yes, I know this function calls exit and closes the program”.

Reply:
wow thats neat